setup/src/Magento/Setup/Module/Di/Code/Reader/FileClassScanner.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+/**
+ * Copyright © 2013-2017 Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Setup\Module\Di\Code\Reader;
+
+class FileClassScanner
+{
+    /**
+     * The filename of the file to introspect
+     *
+     * @var string
+     */
+
+    private $filename;
+
+    /**
+     * The list of classes found in the file.
+     *
+     * @var bool
+     */
+
+    private $classNames = false;
+
+    /**
+     * @var array
+     */
+
+    private $tokens;
+
+    /**
+     * Constructor for the file class scanner.  Requires the filename
+     *
+     * @param string $filename
+     */
+
+    public function __construct($filename)
+    {
+        $filename = realpath($filename);
+        if (!file_exists($filename) || !\is_file($filename)) {
+            throw new InvalidFileException(
+                sprintf(
+                    'The file "%s" does not exist or is not a file',
+                    $filename
+                )
+            );
+        }
+        $this->filename = $filename;
+    }
+
+    /**
+     * Retrieves the contents of a file.  Mostly here for Mock injection
+     *
+     * @return string
+     */
+
+    public function getFileContents()
+    {
+        return file_get_contents($this->filename);
+    }
+
+    /**
+     * Extracts the fully qualified class name from a file.  It only searches for the first match and stops looking
+     * as soon as it enters the class definition itself.
+     *
+     * Warnings are suppressed for this method due to a micro-optimization that only really shows up when this logic
+     * is called several millions of times, which can happen quite easily with even moderately sized codebases.
+     *
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
```

Comment:
@maghamed See https://github.com/magento/magento2/pull/8965#issuecomment-289122432 for the explanation by @kschroeder why there might be a performance impact when decomposing this into more methods. How can we proceed here? Should we try to get hard numbers on the difference in performance?